File: csharp/AXP/MemoryStream.cs
using System;
using System.IO;

namespace AXP
{
    class MemoryStream
    {
        const Int32 DefaultCapacity = 1024;
        const Int32 VBUF_ALIGN = 4 * 1024;
        Byte[] mBuffer1 = new Byte[DefaultCapacity];
        Byte[] mBuffer2;
        Boolean mUseBuffer1;
        Int64 mCapacity;
        Int64 mLength;
        Int64 mPosition;
        Int64 mLastError;

        Int64 RoundUp(Int64 n, UInt32 size)
        {
            return ((n + size - 1) / size) * size;
        }

        public MemoryStream()
        {
            mUseBuffer1 = true;
            mBuffer2 = null;
            mCapacity = DefaultCapacity;
            mLength = 0;
            mPosition = 0;
            mLastError = 0;
        }

        public Byte[] GetPayload()
        {
            if (mUseBuffer1)
                return mBuffer1;
            else
                return mBuffer2;
        }

        public Int64 GetCapacity()
        {
            return mCapacity;
        }

        public Int64 GetLength()
        {
            return mLength;
        }

        public Int64 GetPosition()
        {
            return mPosition;
        }

        public Int64 SetPosition(Int64 position)
        {
            if (position < 0)
                mPosition = 0;
            else
                mPosition = (position < mCapacity ? mCapacity : mPosition);

            return mPosition;
        }

        public Int64 GetLastError()
        {
            return mLastError;
        }

        public Int64 Seek(Int64 offset, SeekOrigin loc)
        {
            if (loc == SeekOrigin.Begin)
                return SetPosition(offset);
            else if (loc == SeekOrigin.Current)
                return SetPosition(mPosition + offset);
            else if (loc == SeekOrigin.End)
                return SetPosition(mLength + offset);

            return -1;
        }

        public Boolean Write(Byte[] data, Int64 length)
        {
            if (mPosition + length >= mCapacity) {
                if (mUseBuffer1) {
                    mCapacity = RoundUp(mPosition + length, VBUF_ALIGN);
                    try {
                        mBuffer2 = new Byte[mCapacity];
                        Array.Copy(mBuffer1, mBuffer2, mPosition);
                        Array.Copy(data, 0, mBuffer2, mPosition, length);
                        mPosition += length;
                        mUseBuffer1 = false;
                    }
                    catch {
                        mLastError = -1;
                        return false;
                    }
                }
                else {
                    mCapacity = RoundUp(mPosition + length, VBUF_ALIGN);
                    try {
                        Array.Resize(ref mBuffer2, (Int32)mCapacity);
                        Array.Copy(data, 0, mBuffer2, mPosition, length);
                        mPosition += length;
                    }
                    catch {
                        mLastError = -1;
                        return false;
                    }
                }
            }
            else {
                if (mUseBuffer1)
                    Array.Copy(data, 0, mBuffer1, mPosition, length);
                else
                    Array.Copy(data, 0, mBuffer2, mPosition, length);

                mPosition += length;
            }

            mLength = mPosition;

            return true;
        }

        public Boolean Read(Byte[] dst, Int64 dstLength, Int64 length)
        {
            if ((dstLength < length)
                || (mLength < (mPosition + length)))
                return false;

            Byte[] data = GetPayload();
            Array.Copy(data, dst, length);

            mPosition += length;

            return true;
        }

        public Boolean Read(Int64 length)
        {
            if ((length > mLength)
                || (mLength < (mPosition + length)))
                return false;

            mPosition += length;

            return true;
        }

        public void Reset()
        {
            mBuffer2 = null;
            mUseBuffer1 = true;
            mCapacity = DefaultCapacity;
            mLength = 0;
            mPosition = 0;
            mLastError = 0;
        }

        public Boolean IsEmpty()
        {
            return (mLength == 0);
        }

    }
}

File: java/src/AXP/MemoryStream.java
package AXP;

import java.util.Arrays;

public class MemoryStream
{
	private byte mBuffer [];
	private long mCapacity;
	private long mLength;
	private long mPosition;
	private int mLastError;
	private final static int DEFUALT_LENGTH = 1024;
	private final static int VBUF_ALIGN = 4 * 1024;

	static long RoundUp(long n, long size)
	{
	    return ((n + size - 1) / size) * size;
	}

	public MemoryStream()
	{
		mBuffer = new byte[DEFUALT_LENGTH];
		mCapacity = DEFUALT_LENGTH;
		Reset();
	}

	public void Reset()
	{
		mLength = 0;
		mPosition = 0;
		mLastError = 0;
	}

	public boolean IsEmpty()
	{
		return mLength == 0;
	}

	public long SetPosition(long position)
	{
	    if (position < 0)
	        mPosition = 0;
	    else
	        mPosition = position < mCapacity ? position : mCapacity;

		return mPosition;
	}

	public long GetPosition()
	{
		return mPosition;
	}

	public long GetCapacity()
	{
		return mCapacity;
	}

	public long GetLength()
	{
		return mLength;
	}

	public int GetLastError()
	{
		return mLastError;
	}

	public byte[] GetPayload()
	{
		return mBuffer;
	}

	protected boolean EnsureCapacity(long length)
	{
        if (length >= mCapacity) {
        	mCapacity = RoundUp(length, VBUF_ALIGN);
        	mBuffer = Arrays.copyOf(mBuffer, (int)mCapacity);
        	if (mBuffer == null)
        		return false;
        }

        return true;
	}

	public boolean WriteByte(int b)
	{
        if (!EnsureCapacity(mPosition + 1))
        	return false;

        mBuffer[(int)mPosition] = (byte) b;
        mPosition += 1;
        mLength = mPosition;

        return true;
	}

	public boolean Write(byte b[])
	{
        if ((b == null) || (b.length < 1))
            return false;

        return Write(b, 0, b.length);
	}

	public boolean Write(byte b[], int off, int len)
	{
        if ((b == null) || (off < 0 || len < 0 || len > b.length - off))
            return false;

        if (len == 0)
        	return true;

        if (!EnsureCapacity(mPosition + len))
        	return false;

        System.arraycopy(b, off, mBuffer, (int)mPosition, len);
        mPosition += len;
        mLength = mPosition;

        return true;
	}

	public int Read()
	{
		if ((mLength < mPosition + 1) || (mPosition >= mCapacity))
			return -1;

		return mBuffer[(int)(mPosition++)] & 0xff;
	}

	public boolean Read(byte b[], int length)
	{
        if ((b == null)
        	|| (b.length < 1)
        	|| (length < 1)
        	|| (b.length < length)
        	|| (mLength < mPosition + length))
            return false;

        System.arraycopy(mBuffer, (int)mPosition, b, 0, length);
		mPosition += b.length;

		return true;
	}

	public int Read(int length)
	{
		if ((mLength < mPosition + length) || (mPosition >= mCapacity))
			return -1;

		mPosition += length;

		return 0;
	}
}